jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Jcmd.java

```diff
@@ -0,0 +1,102 @@
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+package openj9.tools.attach.diagnostics.tools;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+import com.sun.tools.attach.VirtualMachineDescriptor;
+import com.sun.tools.attach.spi.AttachProvider;
+
+import openj9.tools.attach.diagnostics.attacher.AttacherDiagnosticsProvider;
+import openj9.tools.attach.diagnostics.base.DiagnosticUtils;
+
+/**
+ * JMap A tool for running diagnostic commands on another Java process
+ *
+ */
+public class Jcmd {
+
+	@SuppressWarnings("nls")
+	private static final String HELPTEXT = "jcmd: run diagnostic command on another Java process%n"
+			+ " Format:%n"
+			+ "    jcmd <vmid> <arguments>%n"
+			+ " Arguments:%n"
+			+ "    <vmid>: Attach API VM ID as shown in jps or other Attach API-based tools.%n"
+			+ "    command[command arguments]: command from the list returned by \"-help\"%n"
+			+ "    -help: print the list of diagnostic commands%n"
+			+ "    -J: supply arguments to the Java VM running jps%n"
+			+ " NOTE: this utility may significantly affect the performance of the target JVM.%n"
+			+ "     The available diagnostic commands is determined by%n"
+			+ "     the target VM and may vary between VMs.%n"
+			;
+	@SuppressWarnings("nls")
+	private static final String[] HELP_OPTIONS = {"-h", "help", "-help", "--help"};
+
+	/**
+	 * @param args Application arguments.  See help text for details.
+	 */
+	public static void main(String[] args) {
+		String command = null;
+		/* An empty argument list is a request for a list of VMs */
+		final String firstArgument = (0 == args.length)? "-l": args[0]; //$NON-NLS-1$
+		if ("-l".equals(firstArgument)) { //$NON-NLS-1$
+			List<AttachProvider> providers = AttachProvider.providers();
+			AttachProvider myProvider = null;
+			if (0 != providers.size()) {
+				myProvider = providers.get(0);
+			}
+			if (null == myProvider) {
+				System.err.println("no attach providers available"); //$NON-NLS-1$
+			} else {
+				for (VirtualMachineDescriptor vmd : myProvider.listVirtualMachines()) {
+					StringBuilder outputBuffer = new StringBuilder(vmd.id());
+					Util.getTargetInformation(myProvider, vmd, false, false, true, outputBuffer);
+					System.out.println(outputBuffer.toString());
+				}
+			}
+		} else if ((args.length == 1) && Arrays.stream(HELP_OPTIONS).anyMatch(s -> s.equals(firstArgument))) {	
+			System.out.printf(HELPTEXT);
+		} else {
+			String vmid = firstArgument;
+			AttacherDiagnosticsProvider diagProvider = new AttacherDiagnosticsProvider();
+			command = DiagnosticUtils.makeJcmdCommand(args, 1);
+			try {
+				diagProvider.attach(vmid);
```

Comment:
`attach` and `runCommandAndPrintResult` can throw an IOE. `detach` is a null operation if the attach fails. I structured it this way to avoid redundant calls to `handleCommandException`, but  I'm open to suggestions for structuring this better.

Reply:
You can wrap the `runCommandAndPrintResult()` call with a `try` whose `finally` calls `detach()`.

Reply:
Much nicer.  Thanks.